src/opentimelineio/imageSequenceReference.cpp

```diff
@@ -0,0 +1,136 @@
+    _missing_frame_policy {missing_frame_policy} {
+    }
+
+    ImageSequenceReference::~ImageSequenceReference() {
+    }
+
+    RationalTime
+    ImageSequenceReference::frame_duration() const {
+        return RationalTime((double)_value_step, _rate);
+    }
+
+    int ImageSequenceReference::number_of_images_in_sequence() const {
+        if (!this->available_range().has_value()) {
+            return 0;
+        }
+        
+        double playback_rate = (_rate / (double)_value_step);
+        int num_frames = this->available_range().value().duration().to_frames(playback_rate);
+        return num_frames;
+    } 
+
+    std::string
+    ImageSequenceReference::target_url_for_image_number(int const image_number, ErrorStatus* error_status) const {
+        if (image_number >= this->number_of_images_in_sequence()) {
+            *error_status = ErrorStatus(ErrorStatus::ILLEGAL_INDEX);
+            return std::string();
+        }
+        const int file_image_num = _start_value + (image_number * _value_step);
+        const bool is_negative = (file_image_num < 0);
+        std::string image_num_string = std::to_string(abs(file_image_num));
+        std::string zero_pad = std::string();
+        if (image_num_string.length() <  _image_number_zero_padding) {
+            zero_pad = std::string(_image_number_zero_padding - image_num_string.length(), '0');
+        }
+
+        std::string sign = std::string();
+        if (is_negative) {
+            sign = "-";
+        }
+
+        std::string out_string = _target_url_base + _name_prefix + sign + zero_pad + image_num_string + _name_suffix;
+        *error_status = ErrorStatus(ErrorStatus::OK);
+        return out_string;
+    }
+
+    RationalTime
+    ImageSequenceReference::presentation_time_for_image_number(int const image_number, ErrorStatus* error_status) const {
+        if (image_number >= this->number_of_images_in_sequence()) {
+            *error_status = ErrorStatus(ErrorStatus::ILLEGAL_INDEX);
+            return RationalTime();
+        }
+
+        auto first_frame_time = this->available_range().value().start_time();
+        auto time_multiplier = TimeTransform(first_frame_time, image_number, -1);
+        return time_multiplier.applied_to(frame_duration());
+    }
+
+    bool ImageSequenceReference::read_from(Reader& reader) {
+        auto result = reader.read("target_url_base", &_target_url_base) &&
+                reader.read("name_prefix", &_name_prefix) &&
+                reader.read("name_suffix", &_name_suffix) &&
+                reader.read("start_value", &_start_value) &&
+                reader.read("value_step", &_value_step) &&
+                reader.read("rate", &_rate) &&
+                reader.read("image_number_zero_padding", &_image_number_zero_padding);
+
+                std::string missing_frame_policy_value;
+                result && reader.read("missing_frame_policy", &missing_frame_policy_value);
```

Comment:
The return value of read indicates either a key didn't exist, or that the type of the value is unexpected. Was it decided missing_frame_policy should be optional and default to error, or that it must exist in the record? If it's optional, we would need to test the key existence independently of whether it has the correct type.

Reply:
My thought was that it would exist in the record. Since this is new schema we don't have to worry about any previous records of this type.